### PR TITLE
[Model] DiffPool with both DGL and tensorized operatons

### PR DESCRIPTION
**Pull request by HQ01**
## Description

DiffPool Implementation with the first pooling layer computed with DGL operation and the following pooling layer computed with tensorized operation.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by jermainewang** [comment]
Can we add diffpool to NN module first and then use it in model examples? Also, please have a readme about what is this, how to run, results, what each file is for, etc.

**Comment by HQ01** [comment]
Ok, will do.

File: examples/pytorch/diffpool/model/tensorized_layers/diffpool.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from torch import nn as nn
+
+from model.tensorized_layers.assignment import DiffPoolAssignment
+from model.tensorized_layers.graphsage import BatchedGraphSAGE
+
+
+# from model.layers.loss.link_pred import LinkPredLoss
+
+
+class EntropyLoss(nn.Module):
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/model/tensorized_layers/diffpool.py]: Reuse the module above.

File: examples/pytorch/diffpool/model/tensorized_layers/diffpool.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    # Return Scalar
+    def forward(self, adj, anext, s_l):
+        entropy = (torch.distributions.Categorical(probs=s_l).entropy()).sum(-1).mean(-1)
+        assert not torch.isnan(entropy)
+        return entropy
+
+
+class LinkPredLoss(nn.Module):
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/model/tensorized_layers/diffpool.py]: Same

File: examples/pytorch/diffpool/train.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+                        dropout=0.0,
+                        method='diffpool',
+                        bn=True,
+                        bias=True,
+                        save_dir="./model_param",
+                        load_epoch=-1,
+                        data_mode = 'default')
+    return parser.parse_args()
+
+def prepare_data(dataset, prog_args, fold=-1, pre_process=None):
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: Move data processing code to another file

File: examples/pytorch/diffpool/train.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                     hidden_dim, 
+                     embedding_dim,  
+                     label_dim,
+                     activation, 
+                     prog_args.gc_per_block,  
+                     prog_args.dropout, 
+                     prog_args.num_pool, 
+                     prog_args.linkpred,
+                     prog_args.batch_size, 
+                     'maxpool',
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: 'maxpool'?

File: python/dgl/data/tu.py [hunk 5]
```diff
@@ -21,13 +22,17 @@ class TUDataset(object):
 
     """
 
-    _url = r"https://ls11-www.cs.uni-dortmund.de/people/morris/graphkerneldatasets/{}.zip"
+    _url = r"https://ls11-www.cs.tu-dortmund.de/people/morris/graphkerneldatasets/{}.zip"
 
-    def __init__(self, name, use_pandas=False, hidden_size=10):
+    def __init__(self, name, use_pandas=False, hidden_size=10, n_split=0, split_ratio=None, max_allow_node=1000, cross_valid=False):
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: Do not make max_allow_node=1000 as  default. Unless it is specified, otherwise should not remove graphs from dataset.

File: python/dgl/data/tu.py [hunk 6]
```diff
@@ -87,10 +137,17 @@ def __getitem__(self, idx):
...
         """
+        if self.fold != -1:
+            idx += self.fold_start_idx[self.fold]
         g = self.graph_lists[idx]
         return g, self.graph_labels[idx]
 
     def __len__(self):
+        if self.fold != -1:
+            if self.cross_valid:
+                print("WIP")
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: Will we implement this?

File: python/dgl/data/tu.py [hunk 7]
```diff
@@ -60,20 +70,60 @@ def __init__(self, name, use_pandas=False, hidden_size=10):
...
+        #make dataset divisible by 10
+        if len(self.graph_lists) % 10 != 0:
+            print("TU dataset not divisible by 10")
+            new_len = len(self.graph_lists) // 10
+            self.graph_lists = self.graph_lists[:new_len*10]
+            self.graph_labels = self.graph_labels[:new_len*10]
+
+        #check dataset splitting:
+        eps = 1e-5
+        if split_ratio:
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: Since you used Dataloader, I suggest move split and shuffle to the dataloader instead of dataset. The role of Dataset class should mainly be provide sample and corresponding label based on index (implementing `__getitem__` function). Shuffle and split should be done at Dataloader. You can refer to my implementation at https://github.com/VoVAllen/diffpool/blob/master/batched_train.py#L45 (don't follow the practice of moving data to gpu in collate function if you are using num_workers)

File: python/dgl/data/tu.py [hunk 8]
```diff
@@ -60,20 +70,60 @@ def __init__(self, name, use_pandas=False, hidden_size=10):
...
+            assert abs(sum(split_ratio) - 1) < eps, "sum of split ratio is not 1"
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: For KFold, you can write a utils function/class to do this. Such as a custom sampler (https://pytorch.org/docs/stable/_modules/torch/utils/data/sampler.html).

File: examples/pytorch/diffpool/train.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+            train_accu += correct
+            loss = model.loss(ypred, graph_labels)
+            loss.backward()
+            batch_compute_time = time.time() - compute_start
+            computation_time += batch_compute_time
+            nn.utils.clip_grad_norm_(model.parameters(), prog_args.clip)
+            optimizer.step()
+
+
+        train_accu = train_accu / (len(dataloader)*prog_args.batch_size)
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: Change denominator to a `train_count` counter variable. Calculation of number of samples sometimes can be wrong.

File: examples/pytorch/diffpool/train.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+                    batch_graph.ndata[key] = value.cuda()
+                graph_labels = graph_labels.cuda()
+
+                
+            model.zero_grad()
+            compute_start = time.time()
+            ypred = model(batch_graph)
+            indi = torch.argmax(ypred, dim=1)
+            correct = torch.sum(indi == graph_labels).item()
+            train_accu += correct
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: change name to train_true_sample or something else. 
Add `train_count += correct.shape[0]`

File: examples/pytorch/diffpool/README.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+Dependencies
+------------
+* PyTorch 1.0+
+
+How to run
+----------
+
+```bash
+python train.py --dataset ENZYMES --pool_ratio 0.10 --linkpred True --num_pool 1
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/README.md]: Also add sample code to run DD. And describe the corresponding network structure for both dataset.

File: examples/pytorch/diffpool/model/encoder.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            self.pred_input_dim = embedding_dim*self.num_aggs
+        self.pred_layer = nn.Linear(self.pred_input_dim, label_dim)
+
+        # weight initialization
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                m.weight.data = init.xavier_uniform_(m.weight.data,
+                                                    gain=nn.init.calculate_gain('relu'))
+                if m.bias is not None:
+                    m.bias.data = init.constant_(m.bias.data, 0.0)
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/model/encoder.py]: add blank line between functions

File: examples/pytorch/diffpool/train.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+    parser.add_argument('--bn', dest='bn', action='store_const', const=True,
+                        default=True, help='switch for bn')
+    parser.add_argument('--dropout', dest='dropout', type=float,
+                        help='dropout rate')
+    parser.add_argument('--bias', dest='bias', action='store_const',
+                        const=True, default=True, help='switch for bias')
+    parser.add_argument('--save_dir', dest='save_dir', help='model saving directory: SAVE_DICT/DATASET')
+    parser.add_argument('--load_epoch', dest='load_epoch', help='load trained model params from\
+                         SAVE_DICT/DATASET/model-LOAD_EPOCH')
+    parser.add_argument('--data_mode', dest='data_mode', help='data preprocessing mode')
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: Using if statement to determine the data_mode by args.dataset. Doesn't have to be passed by user.

File: examples/pytorch/diffpool/train.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+from model.encoder import DiffPool
+
+def arg_parse():
+    '''
+    argument parser
+    '''
+    parser = argparse.ArgumentParser(description='DiffPool arguments')
+    parser.add_argument('--dataset', dest='dataset', help='Input Dataset')
+    parser.add_argument('--bmname', dest='benchmark name', help='Name of the\
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: What's this

File: examples/pytorch/diffpool/train.py [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+                        benchmark dataset')
+    parser.add_argument('--pool_ratio', dest='pool_ratio', type=float, help='pooling ratio')
+    parser.add_argument('--num_pool', dest='num_pool', type=int, help='num_pooling layer')
+    parser.add_argument('--link_pred', dest='linkpred', action='store_const',
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: change to `store_false` or change to `--no-link-pred` with `store_true`

File: python/dgl/data/tu.py [hunk 16]
```diff
@@ -60,20 +70,60 @@ def __init__(self, name, use_pandas=False, hidden_size=10):
...
+        
+        # randomly shuffle the data
+        ind = [i for i in range(len(self.graph_labels))]
+        random.seed(321)
+        random.shuffle(ind)
+        self.graph_lists = [self.graph_lists[i] for i in ind]
+        self.graph_labels = [self.graph_labels[i] for i in ind]
+
+        #make dataset divisible by 10
+        if len(self.graph_lists) % 10 != 0:
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: Does this have to be divisible by 10?

File: examples/pytorch/diffpool/train.py [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+    parser.add_argument('--batch-size', dest='batch_size', type=int, help='batch size')
+    parser.add_argument('--epochs', dest='epoch', type=int,
+                        help='num-of-epoch')
+    parser.add_argument('--train-ratio', dest='train_ratio', type=float,
+                        help='ratio of trainning dataset split')
+    parser.add_argument('--test-ratio', dest='test_ratio', type=float,
+                        help='ratio of testing dataset split')
+    parser.add_argument('--num_workers', dest='n_worker', type=int,
+                        help='number of workers when dataloading')
+    parser.add_argument('--feature', dest='feature_type',
```
Comment (VoVAllen) [inline examples/pytorch/diffpool/train.py]: Same as data_mode.

File: python/dgl/data/tu.py [hunk 18]
```diff
@@ -60,20 +68,36 @@ def __init__(self, name, use_pandas=False, hidden_size=10):
...
         except IOError:
             print("No Node Attribute Data")
 
         if 'feat' not in g.ndata.keys():
             for idxs, g in zip(node_idx_list, self.graph_lists):
                 g.ndata['feat'] = np.ones((g.number_of_nodes(), hidden_size))
+            self.data_mode = "constant"
             print("Use Constant one as Feature with hidden size {}".format(hidden_size))
+        
+        # remove graphs that are too large by user given standard
```
Comment (VoVAllen) [inline python/dgl/data/tu.py]: Add explanation that this is consistent with authors code
